core(legacy-javascript): new audit

connorjclark opened the pull request:
(previous PR #6730 has too many comments, so here's a new one).

# New Audit: `legacy-javascript` 

**Goal: Encourage developers to serve modern code to modern browsers.**

[Design Doc](https://docs.google.com/document/d/1ItjJwAd6e0Ts6yMbvh8TN3BBh_sAd58rYE1whnpuxaA/edit)

This audit fails if legacy code is found in any first party code. The recommended action is to use the [module/nomodule pattern](https://philipwalton.com/articles/deploying-es2015-code-in-production-today/), in addition to [@babel/preset-modules](https://github.com/babel/preset-modules) (instead of @babel/preset-env).

This first pass does not add this audit to the default config.

## Legacy code, defined

Legacy code contains either polyfills or transformed code for features that modern browsers don't need. "modern browser" is defined as a browser that support ES modules. The relevant polyfills / transforms are in this [spreadsheet](https://docs.google.com/spreadsheets/d/1z28Au8wo8-c2UsM2lDVEOJcI3jOkb2c951xEBqzBKCc/edit).

Polyfills overwrite native prototypes in regular patterns (`String.prototype.startsWith = ...`, `Object.defineProperty(String.prototype, 'startsWith', ...`) or via common libraries (core-js). The patterns that emerge are fairly detectable with a [regex](https://imgflip.com/tag/regex). The major focus was on detecting core-js polyfills, although some effort was made to detect simple ways of polyfilling.

Only some of the transforms are detectable. I went with 3 that were simple and common.

## Woah, [regex](https://gist.githubusercontent.com/connorjclark/542c7ed3727e28cb4ad6ba9bc9109bc2/raw/311a1752a71ffdc09a09e7c4a973c577d78914fb/gistfile1.js). Does it even work?

I wrote a tool to create a bunch of babel projects, varying on individual polyfills / transforms. The results of the tool are also checked into git. All that takes too long to run in CI, so it's just there as a manual verification. Any updates to the audit should also run that tool and check for changes in precision.

## See it

http://misc-hoten.surge.sh/lh-legacy-javascript-pr/www.wikipedia.org_2020-02-06_13-20-38.report.html
http://misc-hoten.surge.sh/lh-legacy-javascript-pr/www.wix.com_2020-02-06_13-27-56.report.html

## Follow up work

1. Fix layout for subRows
1. Use third-party-web
1. Use source maps
1. web.dev article
1. Enable by default
1. Estimate savings to make an opportunity

## Runtime cost

```sh
node lighthouse-core/scripts/compare-timings.js --name my-collection --collect -n 3 --lh-flags='--only-audits=legacy-javascript' --urls https://www.coursehero.com/sg https://www.wix.com https://www.wikipedia.org
node lighthouse-core/scripts/compare-timings.js --name my-collection --summarize --measure-filter 'legacy-javascript'
```

![image](https://user-images.githubusercontent.com/4071474/73978794-e5e43080-48e1-11ea-92ad-aa1cac823526.png)

connorjclark commented on lighthouse-core/scripts/legacy-javascript/summary-sizes.txt:
```diff
@@ -0,0 +1,249 @@
+  29514 variants/core-js-2-only-polyfill/es6-map/main.bundle.min.js
+  29420 variants/core-js-2-only-polyfill/es6-set/main.bundle.min.js
+  27277 variants/core-js-2-only-polyfill/es6-weak-map/main.bundle.min.js
+  24645 variants/core-js-2-only-polyfill/es6-promise/main.bundle.min.js
```
you thought promises were bad? look above ...

connorjclark commented on lighthouse-core/scripts/legacy-javascript/summary-sizes.txt:
```diff
@@ -0,0 +1,249 @@
+
+variants/only-plugin
+2770 total
+1152 variants/only-plugin/-babel-plugin-transform-spread/main.bundle.min.js
```
`only-plugin` sizes will be more interesting when I pull in a real app (prox?). will do for opportunity work.

connorjclark commented on lighthouse-core/scripts/legacy-javascript/update-sizes.sh:
```diff
@@ -0,0 +1,7 @@
+  echo "$d"
+  wc -c "$d"/*/main.bundle.min.js | sort -hr
+  echo -e "\n"
+done > summary-sizes.txt
```
bash is wonderful sometimes

patrickhulce replied:
getting this into the readme mix would be great too :D

patrickhulce commented on .eslintignore:
```diff
@@ -12,3 +12,6 @@
 
 !.eslintrc.js
+
+lighthouse-core/scripts/legacy-javascript/main.js
```
why not just `/* eslint-disable */` in it?

patrickhulce commented on lighthouse-core/audits/legacy-javascript.js:
```diff
@@ -0,0 +1,360 @@
+  title: 'Legacy JavaScript',
+  // eslint-disable-next-line max-len
+  /** Description of a Lighthouse audit that tells the user about old JavaScript that is no longer needed. This is displayed after a user expands the section to see more. No character length limits. 'Learn More' becomes link text to additional documentation. */
+  description: 'Polyfills and transforms enable legacy browsers to use new JavaScript features. However, many aren\'t necessary for modern browsers. For your bundled JavaScript, adopt a modern script deployment strategy using module/nomodule feature detection to reduce the amount of code shipped to modern browsers, while retaining support for legacy browsers. [Learn More](https://philipwalton.com ...
```
I couldn't find a web.dev article on this specifically but there's a neat [codelab](https://web.dev/codelab-serve-modern-code/) that walks you through doing this. a good note for when the docs are written

patrickhulce commented on lighthouse-core/audits/legacy-javascript.js:
```diff
@@ -0,0 +1,360 @@
+}
+
+/**
+ * Identifies polyfills and transforms that should not be present if using module/nomodule pattern.
```
should probably mention your design doc, gist, the scripts dir, and all the other stuff that went into this :)

patrickhulce commented on lighthouse-core/audits/legacy-javascript.js:
```diff
@@ -0,0 +1,360 @@
+    const headings = [
+      /* eslint-disable max-len */
+      {key: 'url', itemType: 'url', subRows: {key: 'locations', itemType: 'source-location'}, text: str_(i18n.UIStrings.columnURL)},
+      {key: '_', itemType: 'code', subRows: {key: 'signals'}, text: ''},
```
is this key unused and it just exists for `subRows`? I'm not sure how to expect this to be rendered, but feels like it deserves a comment either way :)

connorjclark replied:
a _little_ hesitant to add a comment explaining this, since in your IDE you can hover over `key` to get the JSDoc. For a somewhat similar example, we don't explain each usage of `subRows`.

Will give in given a single :-1: :)

patrickhulce replied:
haha fair enough :)

@paulirish @exterkamp am I off base here in my trouble understanding the heading?

patrickhulce commented on lighthouse-core/audits/legacy-javascript.js:
```diff
@@ -0,0 +1,360 @@
+    // Only fail if first party code has legacy code.
+    // TODO(cjamcl): Use third-party-web.
+    const foundSignalInFirstPartyCode = tableRows.some(row => {
+      return row.signals.length && URL.rootDomainsMatch(row.url, mainResource.url);
```
I think this is a decent first pass for 3p filters 👍 

why would there be an entry in here that has no signals though?

also might be worth just filtering them out completely rather than just affecting the passing bit? if they're failing it's going to be annoying to see 90% unactionable things and just ignore the advice because they think they can't pass.

connorjclark replied:
Agreed - that's why I made the audit hide 3p by default in the report, for this audit. I don't much like the idea of _hiding_ insights - this audit should be able to tell you every bit of crufty JS on the page, for those inclined to know. For the rest (majority), hiding 3p by default prevents information overload.

patrickhulce commented on lighthouse-core/config/default-config.js:
```diff
@@ -419,6 +421,8 @@ const defaultConfig = {
         {id: 'resource-summary', weight: 0, group: 'diagnostics'},
         {id: 'third-party-summary', weight: 0, group: 'diagnostics'},
+        // Not enabled yet.
+        // {id: 'legacy-javascript', weight: 0, group: 'diagnostics'},
```
not even in here but without a group?

connorjclark replied:
what do you mean? the audit isn't ready but when it is this line will be uncommented

connorjclark replied:
i'm wondering if #10311 really should have been a "next-config" or "experimental-config" ...

patrickhulce commented on lighthouse-core/audits/legacy-javascript.js:
```diff
@@ -0,0 +1,360 @@
+      return row.signals.length && URL.rootDomainsMatch(row.url, mainResource.url);
+    });
+    return {
+      score: Number(!foundSignalInFirstPartyCode),
```
```suggestion
      score: foundSignalInFirstPartyCode ? 0 : 1,
```

:)

patrickhulce commented on lighthouse-core/lib/i18n/en-US.json:
```diff
@@ -0,0 +1,1866 @@
+{
```
remove this file?

patrickhulce commented on lighthouse-core/report/html/renderer/details-renderer.js:
```diff
@@ -374,10 +374,16 @@ class DetailsRenderer {
-        const valueElement =
-          value !== undefined && !Array.isArray(value) && this._renderTableValue(value, heading);
-        if (valueElement) valueFragment.appendChild(valueElement);
+        if (heading.key === '_') {
```
oh I see this is a new thing! can we make this more explicit like 

```suggestion
        if (heading.key === '__empty-element-with-subrows-only__') {
```

not that literally but something that more obviously suggests what's going on?

connorjclark replied:
what about just `null`?

connorjclark replied:
warnings would get caught in a report test, right?

patrickhulce commented on lighthouse-core/report/html/renderer/report-ui-features.js:
```diff
@@ -268,9 +273,16 @@ class ReportUIFeatures {
       if (!tableEl.parentNode) return; // Keep tsc happy.
       tableEl.parentNode.insertBefore(filterTemplate, tableEl);
+
+      // Hide third-party rows for some audits by default.
```
Why aren't we just hiding these at the audit level?

If it's just for JSON usage I'd prefer personally prefer to stick that into extendedInfo or debugdata or something. The contract we'd extended to consumers of the LHR so far is that items in `.details` are meant to be rendered.

connorjclark replied:
yea you're right. This isn't normally possible, b/c we disable the checkbox by default if there are only all of 1p/3p.

connorjclark replied:
> The contract we'd extended to consumers of the LHR so far is that items in .details are meant to be rendered.

Sorry. I failed to address this.

`extendedInfo` doesn't persist to the LHR, so I don't like that.

How would I add a debugdata in addition to the table details?

Can this be done in a follow up (and be a blocker for becoming enabled by default)?

patrickhulce replied:
Yeah sure, not a blocker for now

patrickhulce commented on lighthouse-core/scripts/legacy-javascript/README.md:
````diff
@@ -0,0 +1,8 @@
+node run.js
+```
+
+Digging into core-js: https://gist.github.com/connorjclark/cc583554ff07cba7cdc416c06721fd6a
````
Could we get a little more explanation of what this directory does and how to use it? :)

I'm thinking... 
- What do the stats in summary-signals and summary-stats represent
- how do we interpret them
- do they need to be updated and asserted like lantern snapshots or sample json
- what would future improvements be doing (adding new core-js pollyfills and testing out new signals? finding real world bundles and checking them in?)
- what does regression/improvement look like when someone does that

connorjclark replied:
> do they need to be updated and asserted like lantern snapshots or sample json

In CI? I don't think this is necessary, but we could do it. `yarn test-legacy-javascript` just before the `git diff --exit-code`, and skip if no files in `scripts/legacy-javascript` / `audits/legacy-javascript.js` were changed.

connorjclark replied:
Right - this audit is checking for the detection of particular build processes (babel / corejs), so it's necessary to have a corpus for debugging / improvement. but the corpus is too large for our CI IMO, and the results shouldn't change except for explicit changes to this audit (for which, one would use this tool and update the summaries for review in PR).

we could do a guarded test like lantern, might be cool, wdyt?

patrickhulce replied:
> we could do a guarded test like lantern, might be cool, wdyt?

Yeah SGTM 👍

patrickhulce commented on lighthouse-core/scripts/legacy-javascript/run.js:
```diff
@@ -0,0 +1,313 @@
+const networkRecordsToDevtoolsLog = require('../../test/network-records-to-devtools-log.js');
+const VARIANT_DIR = `${__dirname}/variants`;
+
+// build, audit, all.
```
oh oh another item for the readme :)

patrickhulce commented on lighthouse-core/scripts/legacy-javascript/run.js:
```diff
@@ -0,0 +1,313 @@
+    fs.writeFileSync(`${dir}/index.html`, `<title>${name}</title><script src=main.bundle.min.js>`);
+
+    // Note: No babelrc will make babel a glorified `cp`.
+    execFileSync('yarn', [
```
what's the expected cwd, `lighthouse-core/scripts/legacy-javascript`? let's assert that or set `cwd` appropriately since most scripts assume repo root is `cwd`

patrickhulce commented on lighthouse-core/scripts/legacy-javascript/run.js:
```diff
@@ -0,0 +1,313 @@
+    const jsRequestWillBeSentEvent = devtoolsLogs.find(e =>
+      e.method === 'Network.requestWillBeSent' && e.params.request.url === scriptUrl);
+    if (!jsRequestWillBeSentEvent) throw new Error('jsRequestWillBeSentEvent is undefined');
+    // @ts-ignore
```
what's being ignored here... wouldn't the line about make us safe?

connorjclark replied:
the log event is not narrowed from `find` :/

patrickhulce replied:
ah gotcha 👍 

```suggestion
    // @ts-ignore - the log event is not narrowed to 'Network.requestWillBeSent' even from find
```

patrickhulce commented on tsconfig.json:
```diff
@@ -25,6 +25,8 @@
     "lighthouse-core/test/**/*.js",
     "clients/test/**/*.js",
     "lighthouse-cli/test/fixtures/**/*.js",
+    "lighthouse-core/scripts/legacy-javascript/main.js",
```
feels like this one could just be `// @ts-nocheck`

patrickhulce commented on lighthouse-core/scripts/legacy-javascript/run.js:
```diff
@@ -0,0 +1,321 @@
+        [LegacyJavascript.DEFAULT_PASS]: devtoolsLogs,
+      },
+      ScriptElements: [
+        // @ts-ignore
```
```suggestion
        // @ts-ignore - partial ScriptElement excluding unused DOM properties
```

patrickhulce commented on lighthouse-core/audits/legacy-javascript.js:
```diff
@@ -0,0 +1,364 @@
+      },
+      {
+        name: '@babel/plugin-transform-spread',
+        expression: /\.apply\(void 0/.source,
```
this is too broad for us to use :/ is there any other reliable pattern that's more specific?

connorjclark replied:
ah, good catch. I'll take a look.

connorjclark replied:
done! `/\.apply\(void 0,\s?_toConsumableArray/`

connorjclark commented on lighthouse-core/scripts/legacy-javascript/run.js:
```diff
@@ -0,0 +1,321 @@
+  const variants = [];
+  for (const dir of glob.sync('*/*', {cwd: VARIANT_DIR})) {
+    const legacyJavascript = require(`${VARIANT_DIR}/${dir}/legacy-javascript.json`);
+    // @ts-ignore
```
ahhhhh it's b/c `acc,cur` are implicit any .... lame.

connorjclark replied:
fixed by typing the dynamic require.

connorjclark commented:
I found that table to be unreadable in GitHub's restricted width styling. paste the markdown here for a better view: https://dillinger.io/

![image](https://user-images.githubusercontent.com/4071474/74282312-c3348c00-4cd4-11ea-94db-5e5fd712d378.png)

connorjclark commented on lighthouse-core/scripts/legacy-javascript/summary-signals.json:
```diff
@@ -521,7 +521,7 @@
     },
     {
       "name": "core-js-2-only-polyfill/es6-weak-map",
-      "signals": "WeakMap, Map"
```
@patrickhulce nice

connorjclark commented:
Thanks Patrick, those `Map=/Set=/etc` false positives were especially egregious. I've modified the regexs and added some false positive tests.

patrickhulce commented on lighthouse-core/audits/legacy-javascript.js:
```diff
@@ -0,0 +1,370 @@
+      extendedInfo: {
+        signalCount,
+      },
+      details,
```
mark as `notApplicable` if there are no polyfills at all?

connorjclark replied:
what about `notApplicable` if no signals from 1p code?

patrickhulce replied:
even better :)

patrickhulce commented:
Looking through some more of the output of https://github.com/GoogleChrome/lighthouse/pull/10327 I wonder if it's even worth surfacing things like `Date.now`, I mean what's that saving, like 28 characters? (though "summary-sizes" seems to think the polyfill is 7KB.... how!?) `Date.now=()=>+new Date`

connorjclark commented:
> Looking through some more of the output of #10327 I wonder if it's even worth surfacing things like `Date.now`, I mean what's that saving, like 28 characters? (though "summary-sizes" seems to think the polyfill is 7KB.... how!?) `Date.now=()=>+new Date`

I see this audit as about collecting signals that imply a non-ideal configuration of babel. Taken in isolation, any one of these polyfills might be implemented in an incredibly minimal way. If you add the entire corejs overhead, you might get more than you'd expect (Date.now being 7KB).

Is there a legit reason first party code might polyfill Date.now, without it coming from an overzealous babel transpilation? If not, I'd want to keep it, as it's an additional signal.

patrickhulce commented:
> Is there a legit reason first party code might polyfill Date.now, without it coming from an overzealous babel transpilation?

I can't think of a reason *first-party* code that's using a build system would do this, but the noise it generates in third-party view is the reason I bring it up. I think you may have said at one point that this only really applies to files built with a build system so maybe what I'm reaching is the same conclusion that we should ignore all files that don't appear to be module bundles.

connorjclark commented:
I see. Let me explore some sites myself and see how noisy it is. If it is unusually more common than the others, I think it makes sense to ignore `Date.now`.

> I think you may have said at one point that this only really applies to files built with a build system so maybe what I'm reaching is the same conclusion that we should ignore all files that don't appear to be module bundles.

Yeah, I did have initial concerns about this. They came from a desire of wanting this to be "easily" actionable - if the file is a bundle, it is safe to assume they are using a modern toolchain, and adding / tweaking babel isn't a huge ordeal. It would amount to detecting specific bundle bootstraps (Webpack, Parcel, etc.) / presence of a source map comment. IIRC, @paulirish didn't like the idea, I don't remember why.

patrickhulce commented:
@paulirish how do you feel about only surfacing from bundle-like files now? can I convince you?

connorjclark commented:
resolved another false positive found from running on paul's site.

Here's what Wikipedia gives

```
URL: https://www.wikipedia.org/portal/wikipedia.org/assets/js/index-c1cb7f1287.js
Polyfills: 2

Array.prototype.indexOf at 0:1753
indexOf||(Array.prototype.indexOf=function(e,t){var n;if(null==this)throw new TypeError('"

Function.prototype.bind at 0:2580
pe.bind||(Function.prototype.bind=function(e){if("function"!=typeof this)throw new TypeErr
```

Should ignore those too, I think.

connorjclark commented:
https://www.redditstatic.com/mweb2x/vendors~Mweb.15c8b87d53fe9156dd3a.js

ctrl + f "node_modules"

never seen that in production before (full module paths)
![image](https://user-images.githubusercontent.com/4071474/74401577-8c469f00-4dd6-11ea-9834-642b2918f768.png)

patrickhulce commented:
@connorjclark you're planning on adding "waiting4reviewer" with a specific call to action when you feel you're blocked on us again, correct?

connorjclark commented:
> Smidge better UX for the empty table problem when all polyfills are 3p (doesn't need designer or anything maybe it's just that we don't enable the filter by default?)

I think the `notApplicable` change handles this, right?

patrickhulce commented:
> I think the notApplicable change handles this, right?

Is the table still empty without a message once you expand the notapplicable audit? That seems to be an equivalent problem to the passing audit with an empty table.

connorjclark commented:
NVM, I had strange expectations for what notApplicable does :)

fixed the empty table thing. also, noticed that the row counting logic was wrong - it assumed that each row would have only one url item type. that is not true, so I changed it to count `tr`s.

connorjclark commented:
anything else reviewers @patrickhulce @paulirish ?

connorjclark commented on lighthouse-core/audits/legacy-javascript.js:
```diff
@@ -0,0 +1,372 @@
+      score: foundSignalInFirstPartyCode ? 0 : 1,
+      notApplicable: !foundSignalInFirstPartyCode,
+      extendedInfo: {
+        signalCount,
```
should this be numericValue and "unitless"?

patrickhulce replied:
IMO if it isn't immediately clear from the unit + name of the audit what the numericValue represents (CLS, dom-size, opportunity, etc) then it should live in extendedInfo, so I like it as-is

patrickhulce commented:
Aight @paulirish speak now or forever hold your peace :)
